### PR TITLE
Custom socket provider and java language server in sidecar

### PR DESCRIPTION
**Pull request by JPinkney**
Signed-off-by: jpinkney <josh.pinkney@mail.utoronto.ca>

### What does this PR do?
Adds the ability to run jdt.ls in a sidecar and allows custom socket launchers.

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/8874

### CODE REVIEW

File: wsagent/che-core-api-languageserver/src/main/java/org/eclipse/che/api/languageserver/remote/CustomSocketLsLauncherProvider.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
```
Comment (ghost) [inline wsagent/che-core-api-languageserver/src/main/java/org/eclipse/che/api/languageserver/remote/CustomSocketLsLauncherProvider.java]: @JPinkney do we really need this? Maybe I am missing smth, but reading ws config and grabbing servers of a special type has been already implemented by @dkuleshov https://github.com/eclipse/che/blob/master/wsagent/che-core-api-languageserver/src/main/java/org/eclipse/che/api/languageserver/remote/SocketLsLauncherProvider.java Can't we just reuse it?

File: multiuser/api/che-multiuser-api-workspace-activity/pom.xml [hunk 2]
```diff
@@ -46,6 +46,10 @@
             <groupId>org.eclipse.che.core</groupId>
             <artifactId>che-core-api-workspace-shared</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.eclipse.che.multiuser</groupId>
+            <artifactId>che-multiuser-api-resource</artifactId>
```
Comment (skabashnyuk) [inline multiuser/api/che-multiuser-api-workspace-activity/pom.xml]: why do you need this artifact?

File: plugins/plugin-java/che-plugin-java-server/src/main/java/org/eclipse/che/plugin/java/languageserver/JavaSocketLanguageServerLauncher.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+import org.eclipse.che.api.languageserver.remote.CustomSocketLanguageServerLauncher;
+import org.eclipse.che.api.languageserver.remote.SocketLanguageServerLauncher;
+import org.eclipse.che.api.languageserver.service.FileContentAccess;
+import org.eclipse.che.api.languageserver.util.DynamicWrapper;
+import org.eclipse.lsp4j.jsonrpc.Launcher;
+import org.eclipse.lsp4j.services.LanguageClient;
+import org.eclipse.lsp4j.services.LanguageServer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
```
Comment (skabashnyuk) [inline plugins/plugin-java/che-plugin-java-server/src/main/java/org/eclipse/che/plugin/java/languageserver/JavaSocketLanguageServerLauncher.java]: javadoc

File: wsagent/che-core-api-languageserver/src/main/java/org/eclipse/che/api/languageserver/remote/SocketLanguageServerLauncher.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+import java.net.Socket;
+import org.eclipse.che.api.languageserver.exception.LanguageServerException;
+import org.eclipse.che.api.languageserver.launcher.LanguageServerLauncher;
+import org.eclipse.che.api.languageserver.launcher.LaunchingStrategy;
+import org.eclipse.che.api.languageserver.launcher.PerWorkspaceLaunchingStrategy;
+import org.eclipse.che.api.languageserver.registry.LanguageServerDescription;
+import org.eclipse.lsp4j.jsonrpc.Launcher;
+import org.eclipse.lsp4j.services.LanguageClient;
+import org.eclipse.lsp4j.services.LanguageServer;
+
```
Comment (skabashnyuk) [inline wsagent/che-core-api-languageserver/src/main/java/org/eclipse/che/api/languageserver/remote/SocketLanguageServerLauncher.java]: javadoc

File: wsagent/che-core-api-languageserver/src/main/java/org/eclipse/che/api/languageserver/remote/SocketLanguageServerLauncher.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    this.languageServerDescription = languageServerDescription;
+    this.host = host;
+    this.port = port;
+  }
+
+  @Override
+  public LanguageServer launch(String projectPath, LanguageClient client)
+      throws LanguageServerException {
+    try {
+      Socket socket = new Socket(host, port);
```
Comment (skabashnyuk) [inline wsagent/che-core-api-languageserver/src/main/java/org/eclipse/che/api/languageserver/remote/SocketLanguageServerLauncher.java]: are you going to close this Socket on ```@PreDestroy```?

**Comment by eivantsov** [comment]
@JPinkney can you write a short how-to with what one can do to test this PR? I have an image with jdt.ls starting in tcp mode.

Also, why does the server need to have attributes - `serverAttributes.get("custom_launcher").equals("true"))`? Right now, it is enought to have `"type": "lsp"` in server attributes in order for this server to be counted as a remote language server.

File: wsagent/che-core-api-languageserver/src/main/java/org/eclipse/che/api/languageserver/remote/SocketLsLauncherProvider.java [hunk 6]
```diff
@@ -91,69 +85,34 @@ public SocketLsLauncherProvider(
...
-              new SocketLanguageServerLauncher(description, host, port);
-          lslRegistry.put(machineName + serverName, launcher);
-        } catch (URISyntaxException e) {
-          LOG.error("Can't parse server url: {}", serverUrl, e);
-        }
-      }
-    }
+          boolean foundCustomLauncher = false;
+          if (serverAttributes.get("custom_launcher") != null
+              && serverAttributes.get("custom_launcher").equals("true")) {
```
Comment (tolusha) [inline wsagent/che-core-api-languageserver/src/main/java/org/eclipse/che/api/languageserver/remote/SocketLsLauncherProvider.java]: It is better to use here `equalsIgnoreCase`

File: plugins/plugin-java/che-plugin-java-server/src/main/java/org/eclipse/che/plugin/java/languageserver/JavaSocketLanguageServerLauncher.java [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+  @Override
+  public boolean isAbleToLaunch() {
+    return true;
+  }
+
+  @Override
+  public LanguageServer launch(String projectPath, LanguageClient client)
+      throws LanguageServerException {
+    try {
+      Socket socket = new Socket(host, port);
```
Comment (tolusha) [inline plugins/plugin-java/che-plugin-java-server/src/main/java/org/eclipse/che/plugin/java/languageserver/JavaSocketLanguageServerLauncher.java]: Should we close this socket later?

File: ide/che-core-ide-stacks/src/main/resources/stacks.json [hunk 8]
```diff
@@ -2961,4 +3033,4 @@
       ]
     }
   }
-]
+]
```
Comment (tolusha) [inline ide/che-core-ide-stacks/src/main/resources/stacks.json]: Don't forget to add empty line.

**Comment by JPinkney** [comment]
To test the instructions are almost the same as the regular sidecar instructions with a few exceptions.

    1. Change the custom recipe to include che-jdt-ls/java-ls as the image (or your own image)
    2. Set the id in the workspace config as "org.eclipse.che.plugin.java.languageserver",  port as 5555, and languageID as java
    3. Mount /home/user/jdtls as a volume for the java-machine-ls

**Comment by eivantsov** [comment]
> 2. Set the id in the workspace config as "org.eclipse.che.plugin.java.languageserver",  port as 5555, and languageID as java

Can you share an example? Where exactly in a workspace config should I provide it?

**Comment by JPinkney** [comment]
Inside the server like this

```
"servers": {
  "ls": {
    "attributes": {
      "internal": "true",
      "type": "ls",
      "config": "{\"id\":\"org.eclipse.che.plugin.java.languageserver\", \"languageIds\":[java]}"
    },
    "protocol": "tcp",
    "port": "5555"
  }
}
```

**Comment by eivantsov** [comment]
So, server attributes :)

Thanks, I will give it a try,

**Review by tsmaeder — CHANGES_REQUESTED** [review]

**Comment by eivantsov** [comment]
@JPinkney yes, that worked - `org.eclipse.che.plugin.java.languageserver` did the trick. I did not expect that since for other LS I am able to assign random IDs.

**Comment by eivantsov** [comment]
Got this in ws agent logs when expanding the project. Otherwise, simple refactoring, auto completion, go to definition, java doc worked fine.
```
Resource '/web-java-spring' already exists.
org.eclipse.core.internal.resources.ResourceException(/web-java-spring)[374]: java.lang.Exception: Resource '/web-java-spring' already exists.
	at org.eclipse.core.internal.resources.ResourceException.provideStackTrace(ResourceException.java:39)
	at org.eclipse.core.internal.resources.Resource.checkDoesNotExist(Resource.java:306)
	at org.eclipse.core.internal.resources.Resource.checkDoesNotExist(Resource.java:293)
	at org.eclipse.core.internal.resources.Project.assertCreateRequirements(Project.java:53)
	at org.eclipse.core.internal.resources.Project.create(Project.java:264)
	at org.eclipse.core.internal.resources.Project.create(Project.java:248)
	at org.eclipse.m2e.core.internal.project.ProjectConfigurationManager.create(ProjectConfigurationManager.java:995)
	at org.eclipse.core.internal.resources.ResourceException.<init>(ResourceException.java:35)
	at org.eclipse.m2e.core.internal.project.ProjectConfigurationManager$1.call(ProjectConfigurationManager.java:1)
	at org.eclipse.m2e.core.internal.project.ProjectConfigurationManager$1.call(ProjectConfigurationManager.java:162)
	at org.eclipse.m2e.core.internal.embedder.MavenExecutionContext.executeBare(MavenExecutionContext.java:177)
	at org.eclipse.m2e.core.internal.embedder.MavenImpl.execute(MavenImpl.java:1355)
	at org.eclipse.m2e.core.internal.embedder.MavenExecutionContext.execute(MavenExecutionContext.java:151)
	at org.eclipse.m2e.core.internal.project.ProjectConfigurationManager.importProjects(ProjectConfigurationManager.java:143)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:60)
	at org.eclipse.jdt.ls.core.internal.managers.MavenProjectImporter.importToWorkspace(MavenProjectImporter.java:151)
2018-05-16 13:38:12,957[pool-9-thread-1]  [ERROR] [o.e.c.a.l.r.CheLanguageClient 74]    - May 16, 2018 1:37:46 PM Error updating workspace folders
	at org.eclipse.m2e.core.internal.embedder.MavenExecutionContext.execute(MavenExecutionContext.java:99)
2018-05-16 13:38:12,960[kspaceUpdater-0]  [ERROR] [.c.p.j.l.WorkspaceSynchronizer 58]   - Workspace updated. Result code: '0', message: 'Error updating workspace folders'. Added projects: '[file:///projects/spring]', removed projects: '[]'
 
	at org.eclipse.jdt.ls.core.internal.managers.ProjectsManager$3.runInWorkspace(ProjectsManager.java:162)
	at org.eclipse.core.internal.resources.InternalWorkspaceJob.run(InternalWorkspaceJob.java:39)
	at org.eclipse.jdt.ls.core.internal.managers.ProjectsManager.access$1(ProjectsManager.java:126)
	at org.eclipse.jdt.ls.core.internal.managers.ProjectsManager.importProjects(ProjectsManager.java:132)
	at org.eclipse.m2e.core.internal.project.ProjectConfigurationManager.importProjects(ProjectConfigurationManager.java:133)
```

**Review by tsmaeder — APPROVED** [review]

**Pull request merged**